provider/bigquery.go

```diff
@@ -335,17 +342,28 @@ func (q defaultBQQueries) createValuePlaceholderString(columns []TableColumn) st
 	return strings.Join(placeholders, ", ")
 }
 
-func (q defaultBQQueries) newBQOfflineTable(name string, columnType string) string {
+func (q defaultBQQueries) newBQOfflineTableQuery(name string, columnType string) string {
 	return fmt.Sprintf("CREATE TABLE `%s` (entity STRING, value %s, ts TIMESTAMP, insert_ts TIMESTAMP)", q.getTableName(name), columnType)
 }
 
-func (q defaultBQQueries) materializationCreate(tableName string, resultName string) string {
-	query := fmt.Sprintf(
-		"CREATE TABLE `%s` AS (SELECT entity, value, ts, row_number() over(ORDER BY entity) as row_number FROM "+
-			"(SELECT entity, ts, value, row_number() OVER (PARTITION BY entity ORDER BY ts DESC, insert_ts DESC) "+
-			"AS rn FROM `%s`) t WHERE rn=1)", q.getTableName(tableName), q.getTableName(resultName))
+func (q defaultBQQueries) materializationCreate(tableName string, schema ResourceSchema, resourceLocation pl.SQLLocation) string {
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("CREATE OR REPLACE TABLE `%s` AS ", tableName))
 
-	return query
+	tsSelectStmt := fmt.Sprintf("`%s` AS ts", schema.TS)
+	tsOrderByStmt := fmt.Sprintf("ORDER BY `%s` DESC", schema.TS)
+	if schema.TS == "" {
+		tsSelectStmt = "TIMESTAMP_SECONDS(0) AS ts"
+		tsOrderByStmt = ""
+	}
+
+	cteFormat := "WITH OrderedSource AS (SELECT `%s` AS entity, `%s` AS value, %s, ROW_NUMBER() OVER (PARTITION BY `%s` %s) AS rn FROM `%s`) "
+	cteClause := fmt.Sprintf(cteFormat, schema.Entity, schema.Value, tsSelectStmt, schema.Entity, tsOrderByStmt, q.getTableNameFromLocation(resourceLocation))
```

Comment:
Impossible to follow this sort of code, I hate how we use Sprintf all over the place to build this SQL strings